pcr_profile.go

```diff
@@ -179,28 +144,109 @@ func (p *PCRProtectionProfile) AddProfileOR(profiles ...*PCRProtectionProfile) *
 	return p
 }
 
-// computePCRValues computes a list of different PCR value combinations from this PCRProtectionProfile.
-func (p *PCRProtectionProfile) computePCRValues(tpm *tpm2.TPMContext, values pcrValuesList) (pcrValuesList, error) {
-	if len(values) == 0 {
-		values = append(values, make(tpm2.PCRValues))
+// pcrProtectionProfileIterator provides a mechanism to perform a depth first traversal of instructions in a PCRProtectionProfile.
+type pcrProtectionProfileIterator struct {
+	instrs [][]pcrProtectionProfileInstr
+}
+
+func (iter *pcrProtectionProfileIterator) descendInToProfiles(profiles ...*PCRProtectionProfile) {
+	instrs := make([][]pcrProtectionProfileInstr, 0, len(profiles)+len(iter.instrs))
+	for _, p := range profiles {
+		pi := make([]pcrProtectionProfileInstr, len(p.instrs))
+		copy(pi, p.instrs)
+		instrs = append(instrs, pi)
 	}
+	instrs = append(instrs, iter.instrs...)
+	iter.instrs = instrs
+}
 
-	for _, instr := range p.instrs {
-		var err error
-		values, err = instr.apply(tpm, values)
-		if err != nil {
-			return nil, err
+func (iter *pcrProtectionProfileIterator) next() pcrProtectionProfileInstr {
+	if len(iter.instrs) == 0 {
+		panic("no more instructions")
+	}
+	if len(iter.instrs[0]) == 0 {
+		iter.instrs = iter.instrs[1:]
+		return &pcrProtectionProfileEndProfileInstr{}
+	}
+
+	instr := iter.instrs[0][0]
+	iter.instrs[0] = iter.instrs[0][1:]
+
+	switch i := instr.(type) {
+	case *pcrProtectionProfileAddProfileORInstr:
+		iter.descendInToProfiles(i.profiles...)
+	}
+
+	return instr
+}
+
+func (iter *pcrProtectionProfileIterator) hasMore() bool {
+	return len(iter.instrs) > 0
+}
+
+// traverseInstructions returns an iterator that performs a depth first traversal through the instructions in this profile.
+func (p *PCRProtectionProfile) traverseInstructions() *pcrProtectionProfileIterator {
+	i := &pcrProtectionProfileIterator{}
+	i.descendInToProfiles(p)
+	return i
+}
+
+// pcrProtectionProfileComputeContext records state used when computing PCR values for a PCRProtectionProfile
+type pcrProtectionProfileComputeContext struct {
+	parent *pcrProtectionProfileComputeContext
+	values pcrValuesList
+}
+
+// computePCRValues computes a list of different PCR value combinations from this PCRProtectionProfile.
+func (p *PCRProtectionProfile) computePCRValues(tpm *tpm2.TPMContext) (pcrValuesList, error) {
+	// Create 2 contexts. The root context is not associated with any profile and exists just to collect PCR values from the root
+	// profile.
```

Comment:
this is confusing, the fact that the root context is not associated with the root profile but collect values from the root profile could use some clarification? it that related that we have a End for it without a starting Or? Maybe having some methods on Context could also help a bit with clarity

Reply:
I've removed this root context, cleaned up the whole of computePCRValues and added some more documentation to hopefully make it a bit more understandable.